File: benchmarks/micro/TailCalls.c
// stark-bench: skip-c-windows
#include <stdint.h>
#include <unistd.h>

#define ITERATIONS UINT64_C(300000)
#define SELF_TEST_ITERATIONS UINT64_C(64)
#define SELF_TEST_SEED UINT64_C(17)
#define SELF_TEST_EXPECTED UINT64_C(318451075583008527)

__attribute__((noinline)) static uint64_t
run_tail_state_machine(uint64_t remaining, uint64_t value) {
  uint8_t state = 0;

  while (remaining != 0) {
    switch (state) {
    case 0:
      value = ((value * UINT64_C(1664525)) + remaining) ^ UINT64_C(1013904223);
      state = 1;
      break;

    case 1:
      value = ((value + (remaining * UINT64_C(1103515245))) ^ (value >> 7)) +
              UINT64_C(12345);
      state = 2;
      break;

    case 2:
      value =
          ((value ^ (remaining + UINT64_C(2654435769))) * UINT64_C(22695477)) +
          UINT64_C(1);
      state = 3;
      break;

    default:
      value =
          (value + (remaining ^ UINT64_C(747796405))) * UINT64_C(2891336453);
      state = 0;
      break;
    }

    remaining -= 1;
  }

  return value;
}

int main(void) {
  if (run_tail_state_machine(SELF_TEST_ITERATIONS, SELF_TEST_SEED) !=
      SELF_TEST_EXPECTED) {
    return 1;
  }

  uint64_t seed = ((uint64_t)getpid()) + SELF_TEST_SEED;
  uint64_t result = run_tail_state_machine(ITERATIONS, seed);
  return (result == 0 || result == seed) ? 1 : 0;
}

File: benchmarks/micro/TailCalls.rs
const ITERATIONS: u64 = 300_000;
const SELF_TEST_ITERATIONS: u64 = 64;
const SELF_TEST_SEED: u64 = 17;
const SELF_TEST_EXPECTED: u64 = 318_451_075_583_008_527;

#[inline(never)]
fn run_tail_state_machine(mut remaining: u64, mut value: u64) -> u64 {
    let mut state = 0_u8;

    while remaining != 0 {
        match state {
            0 => {
                value = value.wrapping_mul(1_664_525).wrapping_add(remaining) ^ 1_013_904_223;
                state = 1;
            }
            1 => {
                value = (value.wrapping_add(remaining.wrapping_mul(1_103_515_245)) ^ (value >> 7))
                    .wrapping_add(12_345);
                state = 2;
            }
            2 => {
                value = (value ^ remaining.wrapping_add(2_654_435_769))
                    .wrapping_mul(22_695_477)
                    .wrapping_add(1);
                state = 3;
            }
            _ => {
                value = value
                    .wrapping_add(remaining ^ 747_796_405)
                    .wrapping_mul(2_891_336_453);
                state = 0;
            }
        }

        remaining -= 1;
    }

    value
}

fn main() {
    if run_tail_state_machine(SELF_TEST_ITERATIONS, SELF_TEST_SEED) != SELF_TEST_EXPECTED {
        std::process::exit(1);
    }

    let seed = (std::process::id() as u64).wrapping_add(SELF_TEST_SEED);
    let result = run_tail_state_machine(ITERATIONS, seed);
    if result == 0 || result == seed {
        std::process::exit(1);
    }
}
